src/main/java/com/hazelcast/aws/AwsDescribeInstancesApi.java

```diff
@@ -113,60 +115,77 @@ private void addFilters(Map<String, String> attributes) {
+            .withReadTimeoutSeconds(awsConfig.getReadTimeoutSeconds())
+            .get();
+        return parse(response);
+    }
 
-        return httpConnection.getInputStream();
+    private static String urlFor(String endpoint, String query) {
+        if (endpoint.startsWith("http")) {
+            return endpoint + "/?" + query;
+        }
+        return "https://" + endpoint + "/?" + query;
     }
 
-    // visible for testing
-    void checkNoAwsErrors(HttpURLConnection httpConnection)
-        throws IOException {
-        int responseCode = httpConnection.getResponseCode();
-        if (isAwsError(responseCode)) {
-            String errorMessage = extractErrorMessage(httpConnection);
-            throw new AwsConnectionException(responseCode, errorMessage);
+    private static Map<String, String> parse(String xmlResponse) {
+        try {
+            return tryParse(xmlResponse);
+        } catch (Exception e) {
+            throw new RuntimeException(e);
         }
     }
 
-    /**
-     * AWS response codes for client and server errors are specified here:
-     * {@see http://docs.aws.amazon.com/AWSEC2/latest/APIReference/errors-overview.html}.
-     */
-    private static boolean isAwsError(int responseCode) {
-        return responseCode >= MIN_HTTP_CODE_FOR_AWS_ERROR && responseCode < MAX_HTTP_CODE_FOR_AWS_ERROR;
+    private static Map<String, String> tryParse(String xmlResponse) throws Exception {
+        return XmlNode.create(xmlResponse)
+            .getSubNodes("reservationset").stream()
+            .flatMap(e -> e.getSubNodes("item").stream())
+            .flatMap(e -> e.getSubNodes("instancesset").stream())
+            .flatMap(e -> e.getSubNodes("item").stream())
+            .filter(e -> e.getValue("privateipaddress") != null)
+            .peek(AwsDescribeInstancesApi::logInstanceName)
+            .collect(Collectors.toMap(
+                e -> e.getValue("privateipaddress"),
+                e -> e.getValue("ipaddress"))
+            );
     }
 
-    private static String extractErrorMessage(HttpURLConnection httpConnection) {
-        InputStream errorStream = httpConnection.getErrorStream();
-        if (errorStream == null) {
-            return "";
-        }
-        return readFrom(errorStream);
+    private static void logInstanceName(XmlNode item) {
+        LOGGER.fine(String.format("Accepting EC2 instance [%s][%s]",
+            parseInstanceName(item),
+            item.getValue("privateipaddress")));
+    }
+
+    private static String parseInstanceName(XmlNode nodeHolder) {
+        return nodeHolder.getSubNodes("tagset").stream()
+            .flatMap(e -> e.getSubNodes("item").stream())
+            .filter(AwsDescribeInstancesApi::isNameField)
+            .flatMap(e -> e.getSubNodes("value").stream())
+            .map(XmlNode::getNode)
+            .map(Node::getFirstChild)
+            .map(Node::getNodeValue)
+            .findFirst()
+            .orElse(null);
```

Comment:
I'm not sure in how many spots this method is used but it'd be probably safer to return `Optional<String>` and let receivers convert to whatever else makes sense for them (even if it's _null_)

Reply:
Changed